iox_catalog/src/sqlite.rs

```diff
@@ -1512,6 +1358,167 @@ WHERE object_store_id = $1;
 
         Ok(Some(parquet_file.into()))
     }
+
+    async fn create_upgrade_delete(
+        &mut self,
+        _partition_id: PartitionId,
+        delete: &[ParquetFile],
+        upgrade: &[ParquetFile],
+        create: &[ParquetFileParams],
+        target_level: CompactionLevel,
+    ) -> Result<Vec<ParquetFileId>> {
+        let mut delete_set = HashSet::new();
+        let mut upgrade_set = HashSet::new();
+        for d in delete {
+            delete_set.insert(d.id.get());
+        }
+        for u in upgrade {
+            upgrade_set.insert(u.id.get());
+        }
+
+        assert!(
+            delete_set.is_disjoint(&upgrade_set),
+            "attempted to upgrade a file scheduled for delete"
+        );
+        let mut tx = self
+            .inner
+            .get_mut()
+            .pool
+            .begin()
+            .await
+            .map_err(|e| Error::StartTransaction { source: e })?;
+
+        let upgrade = upgrade.iter().map(|f| f.id).collect::<Vec<_>>();
+
+        for file in delete {
+            let marked_at = Timestamp::from(self.time_provider.now());
+            flag_for_delete(&mut tx, file.id, marked_at).await?;
+        }
+
+        update_compaction_level(&mut tx, &upgrade, target_level).await?;
+
+        let mut ids = Vec::with_capacity(create.len());
+        for file in create {
+            let res = create_parquet_file(&mut tx, file.clone()).await?;
+            ids.push(res.id);
+        }
+        tx.commit()
+            .await
+            .map_err(|e| Error::FailedToCommit { source: e })?;
+
+        Ok(ids)
+    }
+}
+
+// The following three functions are helpers to the create_upgrade_delete method.
+// They are also used by the respective create/flag_for_delete/update_compaction_level methods.
+async fn create_parquet_file<'q, E>(
+    executor: E,
+    parquet_file_params: ParquetFileParams,
+) -> Result<ParquetFile>
+where
+    E: Executor<'q, Database = Sqlite>,
+{
+    let ParquetFileParams {
+        namespace_id,
+        table_id,
+        partition_id,
+        object_store_id,
+        max_sequence_number,
+        min_time,
+        max_time,
+        file_size_bytes,
+        row_count,
+        compaction_level,
+        created_at,
+        column_set,
+        max_l0_created_at,
+    } = parquet_file_params;
+
+    let query = sqlx::query_as::<_, ParquetFilePod>(
+        r#"
+INSERT INTO parquet_file (
+    shard_id, table_id, partition_id, object_store_id,
+    max_sequence_number, min_time, max_time, file_size_bytes,
+    row_count, compaction_level, created_at, namespace_id, column_set, max_l0_created_at )
+VALUES ( $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14 )
+RETURNING
+    id, table_id, partition_id, object_store_id,
+    max_sequence_number, min_time, max_time, to_delete, file_size_bytes,
+    row_count, compaction_level, created_at, namespace_id, column_set, max_l0_created_at;
```

Comment:
This _could_ be:

```suggestion
RETURNING *;
```

If unless we're purposely selecting columns for the response set?

Reply:
I think the intent is to not return the `shard_id`?

Reply:
Ah yep good spot!